onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,205 @@
+
+  // Operator support related
+ private:
+  int32_t GetMinSupportedNNAPIFeatureLevel(const NodeUnit& node_unit,
+                                           const OpSupportCheckParams& params) const override;
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const NodeUnit& node_unit,
+                         const OpSupportCheckParams& params) const override;
+};
+
+// Add operator related
+
+void ReductionOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  const auto& inputs = node_unit.Inputs();
+  if (inputs.size() > 1 && inputs[1].node_arg.Exists()) {
+    model_builder.AddInitializerToSkip(inputs[1].node_arg.Name());
+  }
+}
+
+Status ReductionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  const auto& op_type(node_unit.OpType());
+  const auto& inputs = node_unit.Inputs();
+  const auto& output = node_unit.Outputs()[0].node_arg.Name();
+
+  auto& shaper(model_builder.GetShaper());
+  const auto input_shape = shaper[inputs[0].node_arg.Name()];
+  const auto& operand_indices(model_builder.GetOperandIndices());
+  const auto& operand_types(model_builder.GetOperandTypes());
+
+  NodeAttrHelper helper(node_unit);
+
+  int32_t op_code;
+  if (op_type == "ReduceMean") {
+    op_code = ANEURALNETWORKS_MEAN;
+  } else {
+    // TODO: Add more reduction ops support
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "ReductionOpBuilder, unknown op: ", op_type);
+  }
+
+  const bool keepdims = helper.Get("keepdims", 1) != 0;
+  const bool noop_with_empty_axes = helper.Get("noop_with_empty_axes", 0) != 0;
+
+  // Get axes for ReduceMean
+  // Note: ONNX `ReduceMean` will reduce by default all dimensions if axes is not provided/provided as empty. However, NNAPI doesn't implement the behavior
+  // to reduce all dimensions by default when 'axes' is empty/not provided. We will convert the case by providing an input with all axes for NNAPI here.
+  // Notes from NNAPI doc:
+  // https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a047fe95a35b27f45c05432b6ca18eb6c
+  std::vector<int32_t> axes;
+  if (node_unit.SinceVersion() >= 18) {
+    if (inputs.size() > 1 && inputs[1].node_arg.Exists()) {
+      // ReduceMean-18 uses the second optional input as axes
+      const auto& initializers(model_builder.GetInitializerTensors());
+      const auto& axes_tensor = *initializers.at(inputs[1].node_arg.Name());
+      Initializer unpacked_tensor(axes_tensor);
+      auto raw_axes = unpacked_tensor.DataAsSpan<int64_t>();
+      axes = OnnxAxesToNnapi(raw_axes, input_shape.size());
+    }
+  } else {
+    // For ReduceMean-13 or earlier, retrieve axes from the attribute
+    const auto& axes_int64 = helper.Get("axes", std::vector<int64_t>{});
+    axes = OnnxAxesToNnapi(gsl::span<const int64_t>(axes_int64.data(), axes_int64.size()), input_shape.size());
```

Comment:
1. assigning the returned vector to a const vector& works (const reference lifetime extension), but it would be clearer to assign to a const vector instead.
see https://abseil.io/tips/107 and https://abseil.io/tips/101 for more info.
2. a `vector<int64_t>` should be convertible to a `gsl::span<const int64_t>`. this flexibility and convenience is a motivation for using span as a function parameter.
```suggestion
    const auto axes_int64 = helper.Get("axes", std::vector<int64_t>{});
    axes = OnnxAxesToNnapi(axes_int64, input_shape.size());
```

Reply:
I see. Thanks for providing the detailed info!